app/uk/gov/hmrc/incometaxsubscriptionselfemployedfrontend/controllers/BusinessNameController.scala

```diff
@@ -79,15 +79,11 @@ class BusinessNameController @Inject()(mcc: MessagesControllerComponents,
             Future.successful(BadRequest(view(formWithErrors, id, isEditMode = isEditMode, isEnabled(SaveAndRetrieve)))),
           businessNameData =>
             multipleSelfEmploymentsService.saveBusinessName(id, businessNameData).map(_ =>
-              if (isEnabled(SaveAndRetrieve)) {
-                Redirect(uk.gov.hmrc.incometaxsubscriptionselfemployedfrontend.controllers.routes.BusinessStartDateController.show(id))
-              }
-              else {
-                if (isEditMode) {
-                  Redirect(uk.gov.hmrc.incometaxsubscriptionselfemployedfrontend.controllers.routes.BusinessListCYAController.show())
-                } else {
-                  Redirect(uk.gov.hmrc.incometaxsubscriptionselfemployedfrontend.controllers.routes.BusinessTradeNameController.show(id))
-                }
+              (isEditMode, isSaveAndRetrieve) match {
+                case (true, true) => uk.gov.hmrc.incometaxsubscriptionselfemployedfrontend.controllers.routes.SelfEmployedCYAController.show(id).url
+                case (false, true) => uk.gov.hmrc.incometaxsubscriptionselfemployedfrontend.controllers.routes.BusinessStartDateController.show(id)
+                case (true, false) => uk.gov.hmrc.incometaxsubscriptionselfemployedfrontend.controllers.routes.BusinessListCYAController.show().url
+                case (false, false) => uk.gov.hmrc.incometaxsubscriptionselfemployedfrontend.controllers.routes.BusinessTradeNameController.show(id).url
```

Comment:
There are few issues with this function:
```
              multipleSelfEmploymentsService.saveBusinessTrade(id, businessTradeNameData).map(_ => {
                val call = (isEditMode, isSaveAndRetrieve) match {
                  case (true, true) => uk.gov.hmrc.incometaxsubscriptionselfemployedfrontend.controllers.routes.SelfEmployedCYAController.show(id)
                  case (false, true) => uk.gov.hmrc.incometaxsubscriptionselfemployedfrontend.controllers.routes.AddressLookupRoutingController.initialiseAddressLookupJourney(id)
                  case (true, false) => uk.gov.hmrc.incometaxsubscriptionselfemployedfrontend.controllers.routes.BusinessListCYAController.show()
                  case (false, false) => uk.gov.hmrc.incometaxsubscriptionselfemployedfrontend.controllers.routes.AddressLookupRoutingController.initialiseAddressLookupJourney(id)
                }
                Redirect(call)
              })
```

Reply:
hi Samir, I think the logic I put is correct. when is not editMode but saveAndRetrieve is enabled, the page after the business name should be the business start date. 
and when it is not in editMode and disabled saveAndRetrieve, the next page from buinsessName should be business trade.

Reply:
Sorry, the logic was wrong in my comment. I added this comment to show you that the pattern matching should return a Call in order to redirect he user.